File: server/src/models/embedded_models/goal_model.py
from mongoengine import *
from src.models.implement_models.user_case_models.user_diet_model import UserDiet
from src.models.implement_models.sample_case_models.sample_diet_model import SampleDiet

class Goal(EmbeddedDocument):
    startWeight = DecimalField(required=True, precision=2, min_value=0)
    targetWeight = DecimalField(required=True, precision=2, min_value=0)
    startPercentBodyFat = IntField(min_value=0, max_value=100)
    targetPercentBodyFat = IntField(min_value=0, max_value=100)
    weightPerWeek = DecimalField(required=True, precision=2, min_value=0)
    targetEnergy = IntField(required=True, min_value=0)
    targetProtein = DecimalField(required=True, precision=2, min_value=0)
    targetFat = DecimalField(required=True, precision=2, min_value=0)
    targetCarbohydrate = DecimalField(required=True, precision=2, min_value=0)
    dietId = ReferenceField(SampleDiet, required=True)

    meta = {
        "db_alias": "miao",
        "collection": "goals"
    }

File: server/src/models/embedded_models/goal.model.js
import mongoose from "mongoose";

const schema = new mongoose.Schema({
  startDate: {
    type: Date,
    required: true,
    default: new Date(),
  },
  startWeight: {
    type: Number,
    required: true,
    default: 0,
    set: (w) => Math.round(w * 10) / 10,
    get: (w) => Math.round(w * 10) / 10,
  },
  targetWeight: {
    type: Number,
    required: true,
    default: 0,
    set: (w) => Math.round(w * 10) / 10,
    get: (w) => Math.round(w * 10) / 10,
  },
  startPercentBodyFat: {
    type: Number,
    default: 0,
    min: 0,
    max: 100,
    set: (f) => Math.round(f * 10) / 10,
    get: (f) => Math.round(f * 10) / 10,
  },
  targetPercentBodyFat: {
    type: Number,
    default: 0,
    min: 0,
    max: 100,
    set: (f) => Math.round(f * 10) / 10,
    get: (f) => Math.round(f * 10) / 10,
  },
  weightPerWeek: {
    type: Number,
    required: true,
    default: 0,
    set: (w) => Math.round(w * 10) / 10,
    get: (w) => Math.round(w * 10) / 10,
  },
  targetEnergy: {
    type: Number,
    required: true,
    default: 0,
    min: 0,
    set: (e) => Math.round(e),
    get: (e) => Math.round(e),
  },
  targetProtein: {
    type: Number,
    required: true,
    default: 0,
    min: 0,
    set: (p) => Math.round(p * 10) / 10,
    get: (p) => Math.round(p * 10) / 10,
  },
  targetFat: {
    type: Number,
    required: true,
    default: 0,
    min: 0,
    set: (f) => Math.round(f * 10) / 10,
    get: (f) => Math.round(f * 10) / 10,
  },
  targetCarbohydrate: {
    type: Number,
    required: true,
    default: 0,
    min: 0,
    set: (c) => Math.round(c * 10) / 10,
    get: (c) => Math.round(c * 10) / 10,
  },
  dietId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "sample_diets",
    required: true,
  },
});

export default schema;
